controllers/Handle.php

```diff
@@ -137,51 +195,38 @@ public function callback(string $provider): RedirectResponse
             'user_type' => get_class($user),
             'user_id' => $user->getKey(),
             'provided_id' => $ssoUser->getId(),
-            'provided_email' => $ssoUser->getEmail(),
+            'provided_email' => $email,
             'ip' => Request::ip(),
             'metadata' => [
                 'remember' => $remember,
             ],
         ]);
 
-        // @TODO: Handle this via an event listener on the backend.auth.login event
-        // and ensure that the event is fired by the AuthManager
-        $runMigrationsOnLogin = (bool) Config::get('cms.runMigrationsOnLogin', Config::get('app.debug', false));
-        if ($runMigrationsOnLogin) {
-            try {
-                // Load version updates
-                UpdateManager::instance()->update();
-            } catch (Exception $e) {
-                Flash::error($e->getMessage());
-            }
-        }
-
-        // @TODO: Also handle via event listener
-        // Log the sign in event
-        AccessLog::add($user);
-
         // Redirect to the intended page after successful sign in
         return Backend::redirectIntended('backend');
     }
 
     /**
      * Redirects the user to the authentication page of the given provider.
+     * Redirects back to signin form on errors with Flash message.
+     * @throws SystemException if the SSO provider is not enabled or is misconfigured.
      */
     public function redirect(string $provider): RedirectResponse
     {
         if (!in_array($provider, $this->enabledProviders)) {
-            return $this->redirectToSigninPage("The {$provider} SSO provider is not enabled");
+            throw new SystemException("The $provider SSO provider is not enabled.");
         }
 
         if ($this->authManager->getUser()) {
-            // @TODO:
-            // - Handle case of user explicitly attaching a SSO provider to their account
-            // - Localization
-            Flash::error('You are already logged in. Please log out first.');
+            // @TODO: Handle case of user explicitly attaching a SSO provider to their account
+            Flash::error(Lang::get('winter.sso::lang.errors.already_logged_in'));
             return Redirect::back();
         }
 
         $config = Config::get('services.' . $provider, []);
+        if (!isset($config['client_id'])) {
+            throw new SystemException("The $provider SSO provider does not have a client_id configured.");
```

Comment:
Same thing here with throwing an Exception.

Reply:
Same answer

Reply:
Here it's different, the provide might just not have been properly configured, even though it's enabled.

Reply:
@LukeTowers how about that for error handling/logging in `redirect()` ?